### PR TITLE
Add TLS support for CQL and JMX

### LINKED ISSUE #129: Enable configuration of secure communication with CQL
Enable configuration of secure communication (TLS/user auth) for the CQL interface in stand-alone application.

### PR DESCRIPTION
**Pull request by emolsson**
Closes #129

### CODE REVIEW

File: connection.impl/src/main/java/com/ericsson/bss/cassandra/ecchronos/connection/impl/LocalJmxConnectionProvider.java [hunk 1]
```diff
@@ -77,15 +84,35 @@ public void close() throws IOException
...
+                else
+                {
+                    System.clearProperty(key);
+                }
+            }
+            env.put("com.sun.jndi.rmi.factory.socket", new SslRMIClientSocketFactory());
+        }
 
-        LOG.debug("Connecting JMX through {}", jmxUrl);
+        LOG.debug("Connecting JMX through {}, credentials: {}, tls: {}", jmxUrl, enabled(credentials != null),
```
Comment (itskarlsson) [inline connection.impl/src/main/java/com/ericsson/bss/cassandra/ecchronos/connection/impl/LocalJmxConnectionProvider.java:114]: I suppose we should change this to true/false if we change the other place. However, here I think this place is more justified with more complex logic.

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java [hunk 2]
```diff
@@ -37,15 +39,25 @@ public DefaultNativeConnectionProvider(Config config, Supplier<Security.CqlSecur
         Config.NativeConnection nativeConfig = config.getConnectionConfig().getCql();
         String host = nativeConfig.getHost();
         int port = nativeConfig.getPort();
-        boolean authEnabled = cqlSecuritySupplier.get().getCredentials().isEnabled();
-        LOG.info("Connecting through CQL using {}:{}, authentication {}", host, port, authEnabled ? "enabled" : "disabled");
+        Security.CqlSecurity cqlSecurity = cqlSecuritySupplier.get();
+        boolean authEnabled = cqlSecurity.getCredentials().isEnabled();
+        boolean tlsEnabled = cqlSecurity.getTls().isEnabled();
+        LOG.info("Connecting through CQL using {}:{}, authentication: {}, tls: {}", host, port, enabled(authEnabled),
```
[the remark this answers is not in the document]
Comment (emolsson) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java:45]: Done

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ReloadingCertificateHandler.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+        tlsConfig.getCipherSuites().ifPresent(sslEngine::setEnabledCipherSuites);
+
+        return new SslHandler(sslEngine);
+    }
+
+    @Override
+    public SslHandler newSSLHandler(SocketChannel channel, InetSocketAddress remoteEndpoint)
+    {
+        throw new UnsupportedOperationException("This method should never be used");
```
[the remark this answers is not in the document]
Comment (emolsson) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ReloadingCertificateHandler.java:84]: Done

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ReloadingCertificateHandler.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    @Override
+    public SslHandler newSSLHandler(SocketChannel channel)
+    {
+        throw new UnsupportedOperationException("This method should never be used");
```
[the remark this answers is not in the document]
Comment (emolsson) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ReloadingCertificateHandler.java:90]: Done

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/TLSConfig.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    private static String[] transformCiphers(String cipher_suites)
+    {
+        if (cipher_suites == null)
+        {
+            return null;
+        }
+
+        return cipher_suites.split(",");
```
[the remark this answers is not in the document]
Comment (emolsson) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/TLSConfig.java:150]: Done

File: docs/SETUP.md [hunk 6]
```diff
@@ -71,24 +71,35 @@ connection:
...
     enabled: true
     username: cassandra
     password: cassandra
 ```
 
-The authentication parameters can be updated during runtime and will automatically be picked up by ecc.
+The security parameters can be updated during runtime and will automatically be picked up by ecc.
 
-If additional properties like SSL is needed it's possible to create custom connection providers.
+It's possible to override the default connection providers and create custom ones.
```
[the remark this answers is not in the document]
Comment (emolsson) [reply docs/SETUP.md:110]: Done

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultJmxConnectionProvider.java [hunk 7]
```diff
@@ -39,9 +43,10 @@ public DefaultJmxConnectionProvider(Config config, Supplier<Security.JmxSecurity
         int port = jmxConfig.getPort();
         LOG.info("Connecting through JMX using {}:{}", host, port);
```
[the remark this answers is not in the document]
Comment (emolsson) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultJmxConnectionProvider.java:44]: Done

File: connection.impl/src/main/java/com/ericsson/bss/cassandra/ecchronos/connection/impl/LocalJmxConnectionProvider.java (hunk 1 above)
Comment (emolsson) [reply connection.impl/src/main/java/com/ericsson/bss/cassandra/ecchronos/connection/impl/LocalJmxConnectionProvider.java:114]: Added booleans outside, so it now it should be quite simple

**Review by itskarlsson — APPROVED** [review]

**Pull request merged**